types/AST.js

```diff
@@ -346,6 +348,8 @@ type CallIndirectInstruction = {
   // WAST
   params?: Array<FuncParam>,
   results?: Array<Valtype>,
+  typeRef?: Identifier | NumberLiteral,
```

Comment:
I'd really like to enforce the constraint that a node has ether `params` & `result` properties, or `typeRef` - is this possible?

Reply:
Yes with a variant type. Maybe something like:

```js
type ParamsAndResults = {
   params?: Array<FuncParam>,
   results?: Array<Valtype>
};

type TypeRef = TypeInstruction | ParamsAndResults

type CallIndirectInstruction = {
  ref: TypeRef
};
```

Reply:
I've expanded on this and outlined in an issue: https://github.com/xtuc/webassemblyjs/issues/257